### ISSUE TITLE
Wrong hyperparameters reflected during optimization

### ISSUE THREAD

**Issue by JoejynWan** [allegroai/clearml#289]
Hi there, 

I am currently using the HyperParameterOptimizer with the following hyperparameters defined in my optimisation script. Arguments are defined using argsparse in the base training task:

```
hyper_parameters=[
    DiscreteParameterRange(
        'Args/epochs', values=[15]),
    DiscreteParameterRange(
        'Args/skip_test', values=[True]),
    DiscreteParameterRange(
        'Args/archi', values=['inception', 'resnet50', 'mobilenet']),
    DiscreteParameterRange(
        'Args/weight_decay', values=[0, 1e-4, 1e-5, 1e-6]),
    UniformParameterRange(
        'Args/lr', min_value=0.0005, max_value=0.005, step_size=0.0005),
    ParameterSet(
        parameter_combinations=[
            {'Args/betadist_alpha':0.9, 'Args/betadist_beta':0.99},
            {'Args/betadist_alpha':0.8, 'Args/betadist_beta':0.9}]),
    DiscreteParameterRange(
        'Args/eps', values=[1e-7, 1e-8, 1e-9]),
    UniformParameterRange(
        'Args/gamma', min_value=0.05, max_value=0.2, step_size=0.05)
],

optimizer_class=OptimizerOptuna,
optuna_pruner=optuna.pruners.HyperbandPruner(),
```

There are a few issues that I have noticed: 
1) Even though the parameters for `epochs` and `archi` are reflected correctly in the CONFIGURATION\HYPER PARAMETERS\Args tab, it is not correctly translated during the running of the training script. Instead, it is using the defaults that were defined in the training script. I think the issue is that integer, string and boolean types are not supported by HyperParameterOptimizer. 

Hyperparameters reflected in the Args tab: 
![image](https://user-images.githubusercontent.com/65271133/104873952-b1e80b00-598c-11eb-9de8-11f185bd12c1.png)

Actual hyperparamters being used, hyperparameters printed from the console: 
![image](https://user-images.githubusercontent.com/65271133/104873894-882ee400-598c-11eb-9bcd-bc2cf4732a8b.png)

2) I am using the Optuna optimiser class, but when I use `ParameterSet`, it gives me the following error, even though I am not using OptimizerBOHB: 
`ValueError: HyperParameter type <class 'clearml.automation.parameters.ParameterSet'> not supported yet with OptimizerBOHB`

3) None of my models are being pruned, so I would just like to make sure that my `optuna_pruner` line is defined correctly. 

Thank you!

**Comment by bmartinn** [comment]
Hi @JoejynWan 
1. All base types should work, en integer example is actually part of the HPO example, as you can see [here](https://github.com/allegroai/clearml/blob/715a0d7edc33b6ff91affe525194fa0c525d98f3/examples/optimization/hyper-parameter-optimization/hyper_parameter_optimizer.py#L72)
That said since these are desecrate options to for the optimizer to choose from, you can check if passing starting will change the behavior in runtime. 
Just making sure, the other parameters were passed correctly ? Is the `clearml-agent` running on the same machine as the HPO process ? 
2.  `ParameterSet` is a way to define a set of configurations for the HPO to choose from for example `adam & lr=0.01` and `SGD & lr=0.001` (this is basically taking discrete parameters and grouping them). Unfortunately Optuna (at least as far as I know) does not support this type of optimization yet, I think BOHB  does support it.
3. Should work (that said no need to pass `optuna_pruner` HyperBand is the default pruner for Optuna)

**Comment by JoejynWan** [comment]
Hi @bmartinn 

1.  Can you elaborate on what you mean by passing starting? 
The other parameters like `lr`, `weight_decay`, `eps` and `gamma` were passed correctly. 
And yes clearml-agent is running on the same machine as the HPO process. Is that not recommended? Currently, my training runs are being completed (all epochs were completed), but it is showing that it failed with error `worker execution exit code 1` (see img below). Is this also the cause of this exit code? 
2. Noted on `ParameterSet` not working for optuna.
3. Noted on HyperBand being the default for optuna. 

![image](https://user-images.githubusercontent.com/65271133/104984874-9b07ee00-5a4a-11eb-954f-a89101cb1ce5.png)

**Comment by bmartinn** [comment]
@JoejynWan this is odd it actually lists the 4 Args parameters (eps,gamma,lr,weight_decay) but not the others ...
1. Could you attach the full log of both the HPO task, and the failed training task?
2. Could you copy-paste from the UI, the HPO Task's Configuration tab, `Configuration Object` / `General` ? 
(this is basically serialization of all the hyper-parameters the HPO will be using)

**Comment by JoejynWan** [comment]
@bmartinn I have attached the logs for the failed training task (dd3e7d9355a44d9b9c9beece22f466d5) and its associated HPO task (c278032e7dd447c99aab622d62eb7fe3). This is the HPO task's configuration tab: 

```
parameter_optimization_space = [{
    type = "DiscreteParameterRange"
    name = "Args/weight_decay"
    values = [0, 0.0001, 1e-05, 1e-06]
  },
  {
    type = "UniformParameterRange"
    name = "Args/lr"
    min_value = 0.0005
    max_value = 0.005
    step_size = 0.0005
    include_max = true
  },
  {
    type = "DiscreteParameterRange"
    name = "Args/eps"
    values = [1e-07, 1e-08, 1e-09]
  },
  {
    type = "UniformParameterRange"
    name = "Args/gamma"
    min_value = 0.05
    max_value = 0.2
    step_size = 0.05
    include_max = true
  }]
```

[task_c278032e7dd447c99aab622d62eb7fe3.log](https://github.com/allegroai/clearml/files/5840368/task_c278032e7dd447c99aab622d62eb7fe3.log)
[task_dd3e7d9355a44d9b9c9beece22f466d5.log](https://github.com/allegroai/clearml/files/5840369/task_dd3e7d9355a44d9b9c9beece22f466d5.log)

Unfortunately, I no longer have the HPO task associated with the task that did not list the parameters correctly (because I accidentally reused the HPO task and it got overwritten). However, I do have the logs for the training task (05ed0e5eef1d4662b111ec5913a598c2) that shows "Using cuda for training with mobilenet architecture.", while in the image below, the args.archi was set to inception. 

![image](https://user-images.githubusercontent.com/65271133/105133096-0de69700-5b27-11eb-9f43-d8a325b19377.png)

[task_05ed0e5eef1d4662b111ec5913a598c2.log](https://github.com/allegroai/clearml/files/5840382/task_05ed0e5eef1d4662b111ec5913a598c2.log)

If required, I can try to replicate the arguments issue and post the logs and HPO Task's Configuration tab.

**Comment by bmartinn** [comment]
@JoejynWan in both cases `task_de403...` and `task_dd3e7` everything looks okay (no crashing not failing)
Also both HPO configuration seem okay (I'm not sure why on one you have `Args/epochs` and not on the other, but I assume this is on purpose)
Do notice that based on the HPO Task log, it seems the default polling is 60 minutes  (which probably is too much if the experiments are short you can change it to poll every minute)

Basically everything seems okay here, what am I missing ?

**Comment by JoejynWan** [comment]
@bmartinn Let's take `task_de403...` as an example. So issues that I am facing here is: 
1) It shows that the training "failed" (the red bar) even though training was completed and the script ran fully. Not sure what this "fail" represents. 
2) In the configuration tab in the UI, it shows that `archi = inception` and that `epochs = 15`. However, in the console (and logs), it is instead training with the mobilenet architecture, and for 1 epoch, which was the defaults in the training script. So the HPO task is giving the training script the correct arguments, but it is not being reflected during the actual training. 

![image](https://user-images.githubusercontent.com/65271133/105360337-702cbe00-5c33-11eb-8dfb-31bf4cd8bf85.png)

![image](https://user-images.githubusercontent.com/65271133/105360424-86d31500-5c33-11eb-8000-7ab0a6d6a7fb.png)

Hope this helps. Let me know if there is an area that needs more clarification. Thanks!

**Comment by bmartinn** [comment]
Thanks @JoejynWan this is what I was looking for.
1. A task is marked failed by the Agent if the exit status is not equal to zero, or it crashed. Can you send the full log of `task_de403...` ? 
2. Yes it seems the arguments were passed but not used by the Task, are those argparse parameters ? how are you using the argparse? Is it the automatically logged ? Can you attach the specific code?
3. Can you test with the latest `clearml` version (from the log I see that the version used is clearml 0.17.2) could you test with `0.17.5rc0` ?

**Comment by JoejynWan** [comment]
Hi @bmartinn, 

1) Yes, the task is marked as failed becuase of `worker execution exit code 1`. May I know what that means? I have previously attached the full logs for `task_de403...`, unless this is not the full logs that you mean? Do let me know if more is required: 
[task_de4036ae02f5420abe64fc26d4a0bdc9.log](https://github.com/allegroai/clearml/files/5878630/task_de4036ae02f5420abe64fc26d4a0bdc9.log)

2) Yes those are argparse parameters. I figured out why the arguments were not used by task. Its because in the training code, the arguments were defined using `parser.parse_args` (e.g. `args = parser.parse_args([ '--archi', 'mobilenet'])`), and so these overwrote the arguments passed by the optimiser. 

3) I have tested it with `0.17.5rc0` and the pattern stated in 2 remained the same. 

Thanks.

**Comment by bmartinn** [comment]
1. Process exit code is the system's way to let you know if the processes finished successfully, or it failed. It seems the process returned 1 (the exit code is something you can control from code, for example `exit(1)` will cause you python code to exit with return code 1). 
Maybe the hint is in the print  "Testing will not be conducted"? Could it be the code exits with return code 1 ?

2. Okay this seems like the issue, you should use `args = parser.parse_args()`, if you want kto have default values you should use the parser default values, make sense ?

**Comment by JoejynWan** [comment]
@bmartinn 

Thank you so much for your help! I used `sys.exit("Testing will not be conducted")`, which I did not know would return with code 1 and cause clearml to determine that has an error. 

And yes I agree that I should use `args = parser.parse_args()`, would solve my problems. Thank you so much for your help!

**Comment by bmartinn** [comment]
This is great news @JoejynWan 
I'm closing this issue, feel free to re-open if we missed something.
